packages/metro/src/DeltaBundler/traverseDependencies.js

```diff
@@ -309,12 +318,107 @@ async function addDependency<T>(
   options.onDependencyAdded();
 }
 
-function removeDependency<T>(
+/**
+ * Recursively look up `inverseDependencies` until it is empty,
+ * returning a set of paths for the last module that does not have
+ * `inverseDependencies`.
+ */
+function getAllTopLevelInverseDependencies<T>(
+  inverseDependencies: Set<string>,
+  graph: Graph<T>,
+  currModule: string,
+  visited: Set<string>,
+): Set<string> {
+  if (visited.has(currModule)) {
+    return new Set();
+  }
+  visited.add(currModule);
+  if (!inverseDependencies.size) {
+    return new Set([currModule]);
+  }
+
+  return Array.from(inverseDependencies)
+    .filter(inverseDep => graph.dependencies.has(inverseDep))
+    .reduce((acc, inverseDep) => {
+      const mod = graph.dependencies.get(inverseDep);
+      if (!mod) {
+        return acc;
+      }
+      getAllTopLevelInverseDependencies(
+        mod.inverseDependencies,
+        graph,
+        inverseDep,
+        visited,
+      ).forEach(x => {
+        acc.add(x);
+      });
+      return acc;
+    }, new Set());
+}
+
+/**
+ * Given `inverseDependencies`, tracing back inverse dependencies to
+ * see if it only leads back to `parentModule`.
+ */
+function canSafelyRemoveFromParentModule<T>(
+  inverseDependencies: Set<string>,
+  parentModule: string,
+  graph: Graph<T>,
+  canBeRemovedSafely: Set<string>,
+  delta: Delta,
+): boolean {
+  const visited = new Set();
+  const result = getAllTopLevelInverseDependencies(
+    inverseDependencies,
+    graph,
+    '',
+    visited,
+  );
+
+  if (!result.size) {
+    /**
+     * This happens when parentModule and inverseDependencies has circular dependency,
+     * this will eventaully became an empty set due to `visited` Set being the base case
+     * for the recursive call.
+     */
+    return true;
+  }
+
+  const filterNotDeletedResult = Array.from(result).filter(
+    x => !delta.deleted.has(x),
+  );
+
+  /**
+   * We can only mark `visited` set of modules to be safely removable if
+   * 1. We do not have top level module to compare with parentModule,
+   *   this can happen when trying to see if we can safely remove from
+   *   a module that was deleted. This is why we filtered them out with `delta.deleted`
+   * 2. We have one top module and it is parentModule
+   *
+   */
+  const canSafelyRemove =
+    !filterNotDeletedResult.length ||
```

Comment:
I think we can remove this line

Reply:
We want to set `canSafelyRemove` when
1. `filterNotDeletedResult` is empty
2. `filterNotDeletedResult` is not empty but only has one item.

I don't think we can remove this 🤔

Reply:
Woops I misread the logic. Definitely do not remove this line 😛